src/image_builder.py

```diff
@@ -154,109 +156,198 @@ def image_builder(buildspec):
             to_build=image_config["build"],
+            stage=constants.INITIAL_STAGE,
             context=context,
         )
 
-        IMAGES.append(image_object)
-
-    FORMATTER.banner("DLC")
-    FORMATTER.title("Status")
-
-    THREADS = {}
-
-    # In the context of the ThreadPoolExecutor each instance of image.build submitted
-    # to it is executed concurrently in a separate thread.
-    with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
-        # Standard images must be built before example images
-        # Example images will use standard images as base
-        standard_images = [image for image in IMAGES if "example" not in image.name.lower()]
-        example_images = [image for image in IMAGES if "example" in image.name.lower()]
-
-        for image in standard_images:
-            THREADS[image.name] = executor.submit(image.build)
+        ##### Create Conclusion stage docker object #####
+        # If for an initial stage image we create a conclusion stage image, then we do not push the initial stage image
+        # to the repository. Instead, we just push its conclusion stage image to the repository. Therefore,
+        # inside function get_conclusion_stage_image_object we make initial_stage_image_object non pushable.
+        conclusion_stage_image_object = get_conclusion_stage_image_object(initial_stage_image_object)
 
-        # the FORMATTER.progress(THREADS) function call also waits until all threads have completed
-        FORMATTER.progress(THREADS)
+        FORMATTER.separator()
 
-        for image in example_images:
-            THREADS[image.name] = executor.submit(image.build)
+        INITIAL_STAGE_IMAGES.append(initial_stage_image_object)
+        if conclusion_stage_image_object is not None:
+            CONCLUSION_STAGE_IMAGES.append(conclusion_stage_image_object)
 
-        # the FORMATTER.progress(THREADS) function call also waits until all threads have completed
-        FORMATTER.progress(THREADS)
+    FORMATTER.banner("DLC")
+    FORMATTER.title("Status")
+    
+    # Standard images must be built before example images
+    # Example images will use standard images as base
+    # Conclusion images must be built at the end as they will consume respective standard and example images
+    standard_images = [image for image in INITIAL_STAGE_IMAGES if "example" not in image.name.lower()]
+    example_images = [image for image in INITIAL_STAGE_IMAGES if "example" in image.name.lower()]
+    conclusion_stage_images = [image for image in CONCLUSION_STAGE_IMAGES]
+    ALL_IMAGES = INITIAL_STAGE_IMAGES + CONCLUSION_STAGE_IMAGES
+    IMAGES_TO_PUSH = [image for image in ALL_IMAGES if image.to_push and image.to_build]
+
+    #initial stage standard images build
+    FORMATTER.banner("Standard Build")
+    build_images(standard_images)
+
+    #initial stage example images build
+    FORMATTER.banner("Example Build")
+    build_images(example_images)
+       
+    #Conclusion stage build
```

Comment:
This stage must run after the `standard_images` to complete the build of standard images. The same must be done after building `example_images`

Reply:
Can we run the CommonStage after the respective builds for standard/example image is complete. This will reduce the build time for the job, especially when there will be more images to build when SM vs DIY is introducted

Reply:
I am not very sure how it reduces the build time for the job. Overall, the number of images being built and pushed will remain the same, right?

Reply:
Lets update the new build process as discussed in the PR description as well

Reply:
Done